### PR TITLE
plan:fix case-when and coalesce type inferer

### PR DESCRIPTION
**Pull request by XuHuaiyu**
fix issue #2911 

refer to MySQL:
https://github.com/mysql/mysql-server/blob/5.7/sql/item_cmpfunc.cc#L4341
https://github.com/mysql/mysql-server/blob/5.7/sql/item_cmpfunc.cc#L4037

PTAL @coocood @shenli @hanfei1991 @tiancaiamao @zimulala

### CODE REVIEW

File: plan/typeinferer.go [hunk 1]
```diff
@@ -561,12 +588,10 @@ func aggArgsType(args []ast.ExprNode) *types.FieldType {
 			unsigned = unsigned && mysql.HasUnsignedFlag(argFieldType.Flag)
 		}
 	}
-	ft := types.NewFieldType(tpClass.ToType())
 	if unsigned {
-		ft.Flag |= mysql.UnsignedFlag
+		*flag |= mysql.UnsignedFlag
 	}
```
Comment (coocood) [inline plan/typeinferer.go]:
```
} else {
  *flag &= ^mysql.UnsignedFlag
}
```

File: plan/typeinferer.go [hunk 2]
```diff
@@ -411,7 +411,22 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 		chs = v.defaultCharset
 		tp.Flen = 40
 	case ast.Coalesce:
+		var gotBinString bool
```
Comment (coocood) [inline plan/typeinferer.go]: This can be computed and returned in `aggClassType`.

File: plan/typeinferer.go [hunk 3]
```diff
@@ -438,37 +453,35 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
...
-		if currType.Tp == mysql.TypeUnspecified {
-			currType = *t
-			continue
-		}
-		mtp := types.MergeFieldType(currType.Tp, t.Tp)
-		if mtp == t.Tp && mtp != currType.Tp {
-			currType.Charset = t.Charset
-			currType.Collate = t.Collate
+	var gotBinString = false
+	exprs := make([]ast.ExprNode, len(x.WhenClauses))
```
Comment (coocood) [inline plan/typeinferer.go]: Need a larger capacity slice for else.

File: plan/typeinferer.go [hunk 4]
```diff
@@ -438,37 +453,35 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
...
+	for i, w := range x.WhenClauses {
+		exprs[i] = w.Result
+		ft := w.Result.GetType()
+		if ft.ToClass() == types.ClassString && mysql.HasBinaryFlag(ft.Flag) {
```
Comment (coocood) [inline plan/typeinferer.go]: We can do it in aggClassType

**Comment by XuHuaiyu** [comment]
PTAL @coocood

File: plan/typeinferer.go [hunk 5]
```diff
@@ -543,15 +538,35 @@ func (v *typeInferrer) convertValueToColumnTypeIfNeeded(x *ast.PatternInExpr) {
...
+		unsigned     bool
+		gotFirst     bool
+		gotBinString bool
+	)
 	for _, arg := range args {
 		argFieldType := arg.GetType()
 		if argFieldType.Tp == mysql.TypeNull {
 			continue
 		}
+		if argFieldType.ToClass() == types.ClassString && mysql.HasBinaryFlag(argFieldType.Flag) {
```
Comment (coocood) [inline plan/typeinferer.go]: `argTypeClass := argFieldType.ToClass()`

Then it can be used multiple times.

File: plan/typeinferer.go [hunk 6]
```diff
@@ -561,12 +576,17 @@ func aggArgsType(args []ast.ExprNode) *types.FieldType {
...
 	if unsigned {
-		ft.Flag |= mysql.UnsignedFlag
+		*flag |= mysql.UnsignedFlag
+	} else {
+		*flag &= ^uint(mysql.UnsignedFlag)
 	}
-	ft.Charset, ft.Collate = types.DefaultCharsetForType(ft.Tp)
-	return ft
+	if tpClass == types.ClassString && !gotBinString {
+		*flag &= ^uint(mysql.BinaryFlag)
```
Comment (coocood) [inline plan/typeinferer.go]: This is repeated work.
Better define a function that turn flag on and off like 
```
func setTypeFlag(*flag uint, flagItem uint, on bool)
```

**Comment by coocood** [comment]
LGTM
@hanfei1991 PTAL

**Pull request merged**